cnxdb/migrations/20170911201035_transform_cnxml_to_html.py

```diff
@@ -0,0 +1,113 @@
+# -*- coding: utf-8 -*-
+import logging
+import time
+from datetime import timedelta
+
+import rhaptos.cnxmlutils
+from dbmigrator import deferred, logger
+
+
+def _batcher(seq, size):
+    for pos in range(0, len(seq), size):
+         yield seq[pos:pos + size]
+
+
+def should_run(cursor, limit='LIMIT 1'):
+    version = rhaptos.cnxmlutils.__version__
+    version_text = '%data-cnxml-to-html-ver="{}"%'.format(version)
+    logger.info('Looking for {}'.format(version_text))
+    cursor.execute("""\
+SELECT files.fileid, max(module_ident)
+    FROM module_files
+    NATURAL JOIN files
+    WHERE filename = 'index.cnxml.html'
+      AND convert_from(file, 'utf-8') NOT LIKE %s
+    GROUP BY files.fileid {}""".format(limit),
+                   (version_text,))
+    return cursor.fetchall()
+
+
+@deferred
+def up(cursor):
+    """Transform all index.cnxml to index.cnxml.html"""
+    # Get all the index.cnxml.html that does not have rhaptos.cnxmlutils
+    # current version
+    to_transform = should_run(cursor, limit='')
+    num_todo = len(to_transform)
+
+    batch_size = 10
+    logger.info('Batch size: {}'.format(batch_size))
+
+    start = time.time()
+    guesstimate = 1 * num_todo
+    guess_complete = guesstimate + start
+    logger.info('Completion guess: '
+                '"{}" ({})'.format(time.ctime(guess_complete),
+                                   timedelta(0, guesstimate)))
+
+    cursor.execute('DROP TABLE IF EXISTS datamigrations.index_cnxml_html')
+
+    # Store module_files for modules we are going to update
+    cursor.execute("""\
+CREATE TABLE datamigrations.index_cnxml_html AS
+    SELECT * FROM module_files
+    WHERE module_ident IN %s
+      AND filename = 'index.cnxml.html'""",
+                   (tuple([i[1] for i in to_transform]),))
+
+    num_complete = 0
+    for batch in _batcher(to_transform, batch_size):
+        module_idents = tuple([i[1] for i in batch])
+        logger.info('Transform module_idents {}'.format(module_idents))
+
+        for fileid, module_ident in batch:
+            cursor.execute("""\
+INSERT INTO files
+    (file, media_type)
+    SELECT convert_to(html_content(%s), 'utf-8'), 'text/xml'
+RETURNING fileid""", (module_ident,))
+            new_fileid = cursor.fetchone()[0]
+            cursor.execute("""\
+UPDATE module_files SET fileid = %s
+WHERE module_ident = %s
```

Comment:
Hmm, how about we make this `WHERE fileid = %oldfileid`, so all the shared usage gets replaced on one go?

Reply:
Oh... yes, this is a bug.  Need to update all the module_files that point to the same index.cnxml.html.  Thanks!